File: Framework/EllipticLicense.m
//
//  EllipticLicense.m
//  EllipticLicense
//
//  Copyright (c) 2009 Dmitry Chestnykh, Coding Robots
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#import "EllipticLicense.h"
#import "NSData+ELAdditions.h"
#include <openssl/err.h>
#include <openssl/evp.h>

@interface EllipticLicense (Private)
- (NSString *)stringSeparatedWithDashes:(NSString *)string;
@end


@implementation EllipticLicense

@synthesize blockedLicenseKeyHashes;

+ (void)initialization;
{
	ERR_load_crypto_strings();
}

- (id)init;
{
	if (![super init])
		return nil;
	[self setCurveName:ELCurveNameSecp112r1]; // Set default curve
	return self;
}

- (id)initWithPublicKey:(NSString *)publicKey privateKey:(NSString *)privateKey curveName:(NSString *)curve;
{
	if (![self init])
		return nil;
	if (curveName)
		[self setCurveName:curve];
	if (![self setPublicKey:publicKey privateKey:privateKey])
		return nil;
	return self;
}

- (id)initWithPublicKey:(NSString *)publicKey privateKey:(NSString *)privateKey;
{
	return [self initWithPublicKey:publicKey privateKey:privateKey curveName:nil];
}

- (id)initWithPublicKey:(NSString *)publicKey curveName:(NSString *)curve;
{
	return [self initWithPublicKey:publicKey privateKey:nil curveName:curve];
}

- (id)initWithPublicKey:(NSString *)publicKey;
{
	return [self initWithPublicKey:publicKey curveName:nil];
}


- (void)setCurveName:(NSString *)name;
{	
	if ([name isEqualToString:ELCurveNameSecp160r1]) {
		curveName = NID_secp160r1;
		digestLength = 20; // Full SHA-1 length
		numberOfDashGroupCharacters = 8;
	}
	else if ([name isEqualToString:ELCurveNameSecp128r1]) {
		curveName = NID_secp128r1;
		digestLength = 16; // SHA-1 is 20 bytes, but since we use 16-byte curve, we must crop it
		numberOfDashGroupCharacters = 4;
	}
	else { // default is ELCurveNameSecp112r1
		curveName = NID_secp112r1;
		digestLength = 14; // SHA-1 is 20 bytes, but since we use 14-byte curve, we must crop it
		numberOfDashGroupCharacters = 5;
	}
}

- (void)setNumberOfDashGroupCharacters:(unsigned int)number;
{
	numberOfDashGroupCharacters = number;
}

- (unsigned int)numberOfDashGroupCharacters;
{
	return numberOfDashGroupCharacters;
}

- (BOOL)setPublicKey:(NSString *)publicKey;
{
	return [self setPublicKey:publicKey privateKey:nil];
}

- (BOOL)setPublicKey:(NSString *)publicKey privateKey:(NSString *)privateKey;
{
	if (ecKey)
		EC_KEY_free(ecKey); 
	ecKey = EC_KEY_new_by_curve_name(curveName);
	if (ecKey == NULL)
		return NO;

	NSData *publicKeyData = [NSData el_dataWithHexString:publicKey];
	const unsigned char *pubBytes = [publicKeyData bytes];
	ecKey = o2i_ECPublicKey(&ecKey, &pubBytes, [publicKeyData length]);
	if (ecKey == NULL)
		return NO;

	if (privateKey) {
		NSData *privateKeyData = [NSData el_dataWithHexString:privateKey];
		const unsigned char *privBytes = [privateKeyData bytes];
		ecKey = d2i_ECPrivateKey(&ecKey, &privBytes, [privateKeyData length]);
		if (ecKey == NULL)
			return NO;
	}
	return (EC_KEY_check_key(ecKey));
}


- (BOOL)generateKeys;
{
	if (ecKey)
		EC_KEY_free(ecKey);
	ecKey = EC_KEY_new_by_curve_name(curveName);
	if (!ecKey)
		return NO;
	return EC_KEY_generate_key(ecKey);
}

- (NSString *)publicKey;
{
	unsigned char *bytes = NULL;
	int length = i2o_ECPublicKey(ecKey, &bytes);
	return [[NSData dataWithBytesNoCopy:bytes length:length] el_hexString];
}

- (NSString *)privateKey;
{
	unsigned char *bytes = NULL;
	int length = i2d_ECPrivateKey(ecKey, &bytes);
	return [[NSData dataWithBytesNoCopy:bytes length:length] el_hexString];
}

- (NSString *)licenseKeyForName:(NSString *)name;
{
	if (!name || [name length] == 0)
		return NO;

    uint8_t digest[digestLength];
    el_compute_digest([name UTF8String], digest, digestLength);

	ECDSA_SIG *signature = ECDSA_do_sign(digest, digestLength, ecKey);
	if (signature == NULL)
		return nil;

	size_t rlen = BN_num_bytes(signature->r);
	size_t slen = BN_num_bytes(signature->s);
	unsigned char *signatureBytes = OPENSSL_malloc(rlen+slen);
	BN_bn2bin(signature->r, signatureBytes);
	BN_bn2bin(signature->s, signatureBytes+rlen); // join two values into signatureBytes
	NSMutableData *signatureData = [NSMutableData dataWithBytesNoCopy:signatureBytes length:rlen+slen];

	return [self stringSeparatedWithDashes:[signatureData el_base32String]];
}

- (BOOL)verifyLicenseKey:(NSString *)licenseKey forName:(NSString *)name;
{
	if (!name || [name length] == 0)
		return NO;

	// Check if license key is blocked. Note that we use key without dashes
	if ([self isBlockedLicenseKey:licenseKey])
		return NO;
	
	ECDSA_SIG *signature = ECDSA_SIG_new();
	if (!signature)
		return NO;
	
	NSData *signatureData = [NSData el_dataWithBase32String:licenseKey];

	// Check length of signature before verifying
	if ([signatureData length] != digestLength * 2)
		return NO;

	size_t partLen = [signatureData length]/2;
	signature->r = BN_bin2bn([signatureData bytes], partLen, signature->r);
	signature->s = BN_bin2bn([signatureData bytes] + partLen, partLen, signature->s);
	if (!signature->r || !signature->s) {
		ECDSA_SIG_free(signature);
		return NO;		
	}

    uint8_t digest[digestLength];
    el_compute_digest([name UTF8String], digest, digestLength);

	BOOL result = ECDSA_do_verify(digest, digestLength, signature, ecKey);

	ECDSA_SIG_free(signature);
	return result;
}

- (NSString *)hashStringOfLicenseKey:(NSString *)licenseKey;
{
    NSData *signatureData = [NSData el_dataWithBase32String:licenseKey];
	return [signatureData el_sha1DigestString];
}

- (BOOL)isBlockedLicenseKey:(NSString *)licenseKey;
{
	if (!blockedLicenseKeyHashes)
		return NO;
	return [blockedLicenseKeyHashes containsObject:[self hashStringOfLicenseKey:licenseKey]];
}

- (void)logKeys;
{
	NSLog(@"Public Key:\n%@", [self publicKey]);
	NSLog(@"Private Key:\n%@", [self privateKey]);
}

- (void)dealloc;
{
	ERR_free_strings();
	if (ecKey)
		EC_KEY_free(ecKey);
}


@end

@implementation EllipticLicense (Private)

- (NSString *)stringSeparatedWithDashes:(NSString *)string;
{
	if (numberOfDashGroupCharacters == 0)
		return [string copy];
	NSMutableString *result = [string mutableCopy];
	int i = numberOfDashGroupCharacters;
	while (i < [result length]) {
		[result insertString:@"-" atIndex:i];
		i += numberOfDashGroupCharacters + 1;
	}	
	return result;
}

@end

File: c_api/elliptic_license.c
//
//  elliptic_license.c
//  EllipticLicense
//
//  Copyright (c) 2013-2017 Vaclav Slavik
//  Copyright (c) 2009 Dmitry Chestnykh, Coding Robots
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.


#include "elliptic_license.h"

#include <string.h>

#ifdef __clang__
// OpenSSL is deprecated in OS X, but still good enough for licensing checks:
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
#endif

#include <openssl/sha.h>
#include <openssl/ec.h>
#include <openssl/ecdsa.h>
#include <openssl/obj_mac.h>
#include <openssl/ossl_typ.h>

#define DIGEST_LENGTH_112   14
#define DIGEST_LENGTH_128   16
#define DIGEST_LENGTH_160   20
#define DIGEST_LENGTH_MAX   DIGEST_LENGTH_160

struct el_context
{
    EC_KEY*        ecKey;
    el_curve_t     curve;
    int            digestLength;
    const uint8_t *blockedKeys;
    int            blockedKeysCount;
};


el_context_t el_create_context(el_curve_t curve,
                               const uint8_t *publicKeyData, int publicKeyLength)
{
    EC_KEY *key = NULL;
    int digestLength = 0;

    switch (curve)
    {
        case el_curve_secp112r1:
            key = EC_KEY_new_by_curve_name(NID_secp112r1);
            digestLength = DIGEST_LENGTH_112;
            break;
        case el_curve_secp128r1:
            key = EC_KEY_new_by_curve_name(NID_secp128r1);
            digestLength = DIGEST_LENGTH_128;
            break;
        case el_curve_secp160r1:
            key = EC_KEY_new_by_curve_name(NID_secp160r1);
            digestLength = DIGEST_LENGTH_160;
            break;
    }

    if (!key)
        return NULL;
    key = o2i_ECPublicKey(&key, &publicKeyData, publicKeyLength);
    if (!key)
        return NULL;

    if (!EC_KEY_check_key(key))
    {
        EC_KEY_free(key);
        return NULL;
    }

    el_context_t ctxt = malloc(sizeof(struct el_context));
    ctxt->ecKey = key;
    ctxt->curve = curve;
    ctxt->digestLength = digestLength;
    ctxt->blockedKeys = NULL;
    ctxt->blockedKeysCount = 0;
    return ctxt;
}


void el_destroy_context(el_context_t ctxt)
{
    if (!ctxt)
        return;
    if (ctxt->ecKey)
        EC_KEY_free(ctxt->ecKey);
    free(ctxt);
}


int el_set_private_key(el_context_t ctxt,
                       const uint8_t *privateKeyData, int privateKeyLength)
{
    ctxt->ecKey = d2i_ECPrivateKey(&ctxt->ecKey, &privateKeyData, privateKeyLength);
    return ctxt->ecKey && EC_KEY_check_key(ctxt->ecKey);
}


void el_set_blocked_keys(el_context_t ctxt,
                         const uint8_t *keyHashes, int dataSize)
{
    ctxt->blockedKeys = keyHashes;
    ctxt->blockedKeysCount = dataSize / SHA_DIGEST_LENGTH;
}


int el_verify_license_key(el_context_t ctxt,
                          const char *licenseKey, const char *name)
{
    // TODO: change this back to use C99 variable length arrays once Visual C++
    //       can deal with it (2013 still can't)
    ECDSA_SIG *signature = NULL;
    uint8_t *signatureData = NULL;
    uint8_t *digest = NULL;

    if (!licenseKey || !strlen(licenseKey) || !name || !strlen(name))
        return 0;

    int signatureLength = el_base32_decode_buffer_size((int)strlen(licenseKey));

    signatureData = malloc(signatureLength);
    signatureLength = el_base32_decode(licenseKey, signatureData, signatureLength);

    // Check length of signature before verifying
    if (signatureLength != ctxt->digestLength * 2)
    {
        free(signatureData);
        return 0;
    }

    signature = ECDSA_SIG_new();
    if (!signature)
    {
        free(signatureData);
        return 0;
    }

    int partLen = signatureLength / 2;
    signature->r = BN_bin2bn(signatureData,           partLen, signature->r);
    signature->s = BN_bin2bn(signatureData + partLen, partLen, signature->s);
    if (!signature->r || !signature->s)
    {
        free(signatureData);
        ECDSA_SIG_free(signature);
        return 0;
    }

    digest = malloc(ctxt->digestLength);
    el_compute_digest(name, digest, ctxt->digestLength);

    int result = ECDSA_do_verify(digest, ctxt->digestLength, signature, ctxt->ecKey) == 1;

    // If the key was accepted, check it against a blacklist of blocked keys:
    if (result && ctxt->blockedKeys)
    {
        unsigned char sha1hash[SHA_DIGEST_LENGTH];
        SHA1(signatureData, signatureLength, sha1hash);
        const uint8_t *ptr = ctxt->blockedKeys;
        for (int i = 0; i < ctxt->blockedKeysCount; i++, ptr += SHA_DIGEST_LENGTH)
        {
            if (memcmp(sha1hash, ptr, SHA_DIGEST_LENGTH) == 0)
            {
                result = 0;
                break;
            }
        }
    }

    free(signatureData);
    free(digest);
    ECDSA_SIG_free(signature);

    return result;
}


int el_generate_license_key(el_context_t ctxt,
                            const char *name, char *output)
{
    int signatureLength = 2 * ctxt->digestLength;
    int bufferLength = el_base32_encode_buffer_size(signatureLength);

    if (output == NULL)
        return bufferLength;

    unsigned char digest[DIGEST_LENGTH_MAX];
    el_compute_digest(name, digest, ctxt->digestLength);

    unsigned char signatureBytes[2 * DIGEST_LENGTH_MAX];

    // The length of ECDSA_SIG's r and s components may be shorter than digestLength rarely
    // (see docs, incl. BN_num_bytes). We want fixed-length license keys, so just discard
    // these results.
    for (;;)
    {
        ECDSA_SIG *sig = ECDSA_do_sign(digest, ctxt->digestLength, ctxt->ecKey);
        if (sig == NULL)
            return -1;
        int rlen = BN_num_bytes(sig->r);
        int slen = BN_num_bytes(sig->s);
        if (rlen + slen == signatureLength)
        {
            BN_bn2bin(sig->r, signatureBytes);
            BN_bn2bin(sig->s, signatureBytes+rlen); // join two values into signatureBytes
            ECDSA_SIG_free(sig);
            break;
        }
        // else: try again
        ECDSA_SIG_free(sig);
    }

    el_base32_encode(signatureBytes, signatureLength, output, bufferLength);
    return bufferLength;
}


void el_compute_digest(const char *name, uint8_t *digest, int digestSize)
{
    uint8_t sha256_digest[SHA256_DIGEST_LENGTH];

    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, name, strlen(name));
    SHA256_Final(sha256_digest, &sha256);

    memcpy(digest, sha256_digest, digestSize);
}
